examples/rcn/inference_pgmax_small.py

```diff
@@ -0,0 +1,203 @@
+# ---
+# jupyter:
+#   jupytext:
+#     formats: ipynb,py:percent
+#     text_representation:
+#       extension: .py
+#       format_name: percent
+#       format_version: '1.3'
+#       jupytext_version: 1.11.4
+#   kernelspec:
+#     display_name: Python 3
+#     language: python
+#     name: python3
+# ---
+
+# %%
+# %matplotlib inline
+import os
+import time
+
+import matplotlib.pyplot as plt
+import numpy as np
+from helpers import (
+    get_number_of_states,
+    index_to_rc,
+    initialize_evidences,
+    visualize_graph,
+)
+from jax import jit
+from load_data import get_mnist_data_iters
+
+from pgmax.fg import graph, groups
+
+os.environ["XLA_PYTHON_CLIENT_ALLOCATOR"] = "platform"
+os.environ["XLA_PYTHON_CLIENT_PREALLOCATE"] = "false"
+
+# # 1. Load the data
+
+hps, vps = 11, 11
+train_size = 20
+test_size = 20
+data_dir = "/storage/users/skushagra/MNIST/"
+
+# +
+train_set, test_set = get_mnist_data_iters(data_dir, train_size, test_size)
+
+train_labels = -1 + np.zeros(len(train_set))
+for i in range(len(train_set)):
+    train_labels[i] = train_set[i][1]
+
+test_labels = -1 + np.zeros(len(test_set))
+for i in range(len(test_set)):
+    test_labels[i] = test_set[i][1]
+# -
+
+# # 2. Load the model
+
+directory = f"model_{train_size}_{hps}_{vps}"
+frcs = np.load(f"{directory}/frcs.npy", allow_pickle=True, encoding="latin1")
+edges = np.load(f"{directory}/edges.npy", allow_pickle=True, encoding="latin1")
+phis = np.load(f"{directory}/phis.npy", allow_pickle=True, encoding="latin1")
+M = get_number_of_states(hps, vps)
+
+img = visualize_graph(frcs[4], edges[4])
+
+# # 3. Make pgmax graph
+
+# ## 3.1 Make variables
+
+# +
+start = time.time()
+assert frcs.shape[0] == edges.shape[0]
+
+variables_all_models = {}
+for idx in range(frcs.shape[0]):
+    frc = frcs[idx]
+    variables_all_models[idx] = groups.NDVariableArray(
+        variable_size=M, shape=(frc.shape[0],)
+    )
+
+end = time.time()
+print(f"Creating variables took {end-start:.3f} seconds.")
+# -
+
+# ## 3.2 Make factors
+
+# +
+start = end
+
+fg = graph.FactorGraph(variables=variables_all_models)
+for idx in range(edges.shape[0]):
+    edge = edges[idx]
+
+    for e in edge:
+        i1, i2, r = e
+        fg.add_factor(
+            [(idx, i1), (idx, i2)],
+            phis[r],
```

Comment:
We can easily construct these `phis` on demand so probably better to construct them instead of loading them from disk.

Reply:
Computing `phi` involves the following code


    rows = []
    cols = []
    index = 0
    for i in range(1, M):
        r1, c1 = index_to_rc(i, hps, vps)

        r2_min = max(r1 - r, -hps)
        r2_max = min(r1 + r, hps)
        c2_min = max(c1 - r, -vps)
        c2_max = min(c1 + r, vps)

        for r2 in range(r2_min, r2_max + 1):
            for c2 in range(c2_min, c2_max + 1):
                j = rc_to_index(r2, c2, hps, vps)
                rows.append(i)
                cols.append(j)
                index += 1

which is not too complicated. For simplicity of the current notebook; I didnt include it in the example. We can include it if you think it helps the presentation.